doc/notes/188_enhance_logging_ant.adoc

```diff
@@ -0,0 +1,369 @@
+format.
+
+==== 4.4.1 Audit Event Statistics
+
+The primary statistics surrounding audit events are focused around timing.
+A logger specifically used for benchmarking shall be defined.  This
+benchmarking logger will log an entry at event arrival and another when
+the event is processed by Sequence Verification.
+
+=== 4.5 Protocol Verifier Statistics
+
+Several parameters are useful in the maintenance and support of the
+Protocol Verifier (itself, as opposed to the application the PV is
+monitoring).
+
+The following statistics are reported as they happen.
+
+* worker registered
+* worker deregistered
+* unassigned job count exceeded
+  ** unassigned job count exceeded is reported once upon occurrence and
+     then only reported again if the unassigned job count goes below the
+     threshold.
+
+The following statistics are reported periodically.
+
+* event count since start
+* event throughput
+* worker count
+* concurrent jobs at each worker
+* concurrent jobs overall (Assigned Jobs)
+* unassigned jobs
+* unassigned job count exceeded
+
+This information shall be reported on a regular basis not exceeding once
+per second (1 Hertz).  The timing shall be configurable.
+
+The recommended location of this logging is in Job Management.
+
+=== 4.6 Logger Hierarchy
+
+As described above, `log4cplus` supports a logger hierarchy.  This
+provides a dimension of configuration that can be quite helpful.  For
+example, if we define loggers at the class-level, by default they can be
+left unconfigured and simply inherit the configuration of the domain
+logger.  And then when selective and isolated debugging is needed
+(especially in a production environment), it can be accomplished class by
+class.
+
+* The top-level logger shall be 'pv' (Protocol Verifier).
+* Each domain shall have a logger.
+  ** 'pv.jm':  Job Management
+  ** 'pv.aeo':  Audit Event Ordering
+  ** 'pv.sv':  Sequence Verification
+  ** 'pv.istore':  Invariant Store
+  ** 'pv.vg':  Verification Gateway
+     *** 'pv.vg.audit' will be the logger channel used to report PVprime
+          audit events.
+* Domain-level debug loggers shall append the class name or keyletters to
+  the domain-level logger designation.  Most of the Logger entries for
+  domain logging will be at the DEBUG level.  However, INFO or WARN level
+  logging from these domain level loggers will go to the Reporting back
+  end.
+  ** 'pv.sv.happyjob' (or 'pv.sv.job') INFO to declare job success, WARN
+     to declare job failure or alarm, DEBUG for engineering
+     instrumentation.
+     *** 'pv.sv.happyjob.assessingcontraints' is an example state machine
+     state of HappyJob.
+  ** 'pv.sv.unhappyjob' ditto
+  ** 'pv.sv.acceptevent' is a domain function in Sequence Verification.
+* A special logger shall be defined for benchmarking.
+  ** 'pv.benchmark':  throughput benchmark (likely used at event reception
+  in Job Management and at event processed in Sequence Verification.
+
```

Comment:
Is this a valid pictorial representation of the words here?
[Logger Hierarchy.pptx](https://github.com/xtuml/munin/files/14722518/Logger.Hierarchy.pptx)

Reply:
Close.
You missed a level of hierarchy.  The max I used was 4 levels.
Hierarchy is independent of level.  Any level can be logged at any depth of the hierarchy.